packages/kit/src/core/sync/write_env.js

```diff
@@ -0,0 +1,183 @@
+export let env = {};
+
+/** @type {(environment: Record<string, string>) => void} */
+export function set_env(environment) {
+  env = environment;
+}`;
+}
+
+/**
+ * Writes the existing environment variables in process.env to
+ * $app/env and $app/env/private
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {string | undefined} mode
+ * The Vite mode.
+ */
+export function write_env(config, mode = undefined) {
+	const pub = write_public_env(config, mode);
+	const priv = write_private_env(config, mode);
+	write_runtime_env(config);
+	write_typedef(config, pub, priv);
+}
+
+/**
+ * Writes the existing environment variables prefixed with config.kit.env.publicPrefix
+ * in process.env to $app/env
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {string | undefined} mode
+ * The Vite mode.
+ */
+function write_public_env(config, mode = undefined) {
+	const pub_out = path.join(config.outDir, 'runtime/app/env.js');
+
+	// public is a little difficult since we append to an
+	// already-existing file
+	const pub = resolve_public_env(config.env.publicPrefix, mode);
+	let pub_content = const_declaration_template(true, pub);
+	if (fs.existsSync(pub_out)) {
+		const old_pub_content = fs.readFileSync(pub_out).toString();
+		const autogen_content_start = old_pub_content.indexOf(autogen_comment);
+		if (autogen_content_start === -1) {
+			pub_content = old_pub_content + '\n' + pub_content;
+		} else {
+			pub_content = old_pub_content.slice(0, autogen_content_start) + '\n' + pub_content;
+		}
+	}
+	write_if_changed(pub_out, pub_content);
+	return pub;
+}
+
+/**
+ * Writes the existing environment variables not prefixed with config.kit.env.publicPrefix
+ * in process.env to $app/env/private
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {string | undefined} mode
+ * The Vite mode.
+ */
+function write_private_env(config, mode = undefined) {
+	const priv_out = path.join(config.outDir, 'runtime/app/env/private.js');
+
+	// private is easy since it has its own file: just write if changed
+	const priv = resolve_private_env(config.env.publicPrefix, mode);
+	const priv_content = const_declaration_template(false, priv);
+	write_if_changed(priv_out, priv_content);
+	return priv;
+}
+
+/**
+ * Writes a blank export to $app/env/runtime. This should be populated by
+ * Server.init whenever an adapter calls it
+ * @param {import('types').ValidatedKitConfig} config
+ */
+function write_runtime_env(config) {
+	const runtime_out = path.join(config.outDir, 'runtime/app/env/runtime.js');
+
+	write_if_changed(runtime_out, runtime_env_template());
+}
+
+/**
+ * Writes the type definitions for the environment variable files
+ * to types/ambient.d.ts
+ * @param {import('types').ValidatedKitConfig} config
+ * @param {Record<string, string>} pub
+ * @param {Record<string, string>} priv
+ */
+function write_typedef(config, pub, priv) {
+	const type_declaration_out = path.join(config.outDir, 'types/ambient.d.ts');
```

Comment:
mostly unrelated to this PR, but this gets me thinking: maybe we should reference the Kit types in this ambient file, rather than here: https://github.com/sveltejs/kit/blob/ab65de0c065df5e274082805bc379186eb6c15d3/packages/create-svelte/templates/default/src/app.d.ts#L1

Tidier, and one less thing to break in userland. Or maybe even that's too complicated and we can reference those types directly in the generated config? not sure

Reply:
I don't think I totally understand what you're suggesting. Can you flesh it out a bit more?

Reply:
Currently, in order to get the ambient module types defined in https://github.com/sveltejs/kit/blob/master/packages/kit/types/ambient.d.ts, you need to add this line to a module that TypeScript can 'see':

```ts
/// <reference types="@sveltejs/kit" /> 
```

The expectation is that you include it in `app.d.ts`, and that's where it is in the templates. But someone might want to organise their types differently, or whatever, and that reference could easily get lost (or they could create a project without types then want to add them later, etc, and this is one more thing they'd need to worry about).

If we referenced the types from the generated tsconfig.json somehow, then they'd no longer need to worry about that (as long as they `extend` the config correctly).

Reply:
opened an issue for this, we can tackle after this PR https://github.com/sveltejs/kit/issues/5677